### PR TITLE
Adding Docker images for transformers + notebooks

### PR DESCRIPTION
**Pull request by mfuntowicz**
Docker images are as follow:

- transformers-cpu (PyTorch + TF)
- transformers-gpu (PyTorch + TF)
- transformers-pytorch-cpu
- transformers-pytorch-gpu
- transformers-tensorflow-cpu
- transformers-tensorflow-gpu

Images are tagged according to the version of the library they bring and always use the latest version of DL frameworks.

Notebooks introduce:

- How to use tokenizers
- Overall transformers overview
- How to use pipelines 

Currently notebooks are added to the repo, let's discuss internally if it's a good idea.

### CODE REVIEW

**Comment by julien-c** [comment]
On the Docker files, do we really need conda? (I try to stay away from conda as much as possible)

**Comment by mfuntowicz** [comment]
Well, conda brings Intel MKL with just `conda install mkl, mkl-devel` which improves PyTorch and TF by a significant factor. 

Depending on what level of performances we want to provide: 

- I totally remove MKL and install some Open BLAS/LAPACK libraries
-I'll build MKL in the images and include into PATH

**Comment by BramVanroy** [comment]
MKL is also available on PyPi. Have a looks-y [here](https://software.intel.com/en-us/articles/installing-the-intel-distribution-for-python-and-intel-performance-libraries-with-pip-and) to check whether everything you need is there.

File: notebooks/02-transformers.ipynb [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+    "\n",
+    "And also, the Attention mechanism, which introduced a good improvement over \"raw\" RNNs by giving \n",
+    "a learned, weighted-importance to each element in the sequence, allowing the model to focus on important elements.\n",
+    "\n",
+    "![attention_rnn](https://3qeqpr26caki16dnhd19sv6by6v-wpengine.netdna-ssl.com/wp-content/uploads/2017/08/Example-of-Attention.png)  \n",
+    "\n",
+    "## Then comes the Transformer  \n",
+    "\n",
+    "The Transformers era originally started from the work of [(Vaswani & al., 2017)](https://arxiv.org/abs/1706.03762) who\n",
+    "demonstrated it's superiority over [Recurrent Neural Network (RNN)](https://en.wikipedia.org/wiki/Recurrent_neural_network)\n",
```
Comment (LysandreJik) [inline notebooks/02-transformers.ipynb]: it's > its

File: notebooks/02-transformers.ipynb [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+    "on translation tasks but it quickly extended to almost all the tasks RNNs were State-of-the-Art at that time.\n",
+    "\n",
+    "One advantage of Transformer over it's RNN counterpart was it's non sequential attention model. Remember, the RNNs had to\n",
```
Comment (LysandreJik) [inline notebooks/02-transformers.ipynb]: it's > its

Comment (LysandreJik) [reply notebooks/02-transformers.ipynb]: x2

File: notebooks/02-transformers.ipynb [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+    "# transformers generates a ready to use dictionary with all the required parameters for the specific framework.\n",
+    "input_tf = tokenizer.encode_plus(\"This is a sample input\", return_tensors=\"tf\")\n",
+    "input_pt = tokenizer.encode_plus(\"This is a sample input\", return_tensors=\"pt\")\n",
+    "\n",
+    "# Let's compare the outputs\n",
+    "output_tf, output_pt = model_tf(input_tf), model_pt(**input_pt)\n",
+    "\n",
+    "# Models outputs 2 values (The value for each tokens, the pooled representation of the input sentence)\n",
+    "for name, o_tf, o_pt in zip([\"output\", \"pooled\"], output_tf, output_pt):\n",
+    "    print(\"{} differences: {}\".format(name, (o_tf.numpy() - o_pt.numpy()).sum()))"
```
Comment (LysandreJik) [inline notebooks/02-transformers.ipynb]: Maybe explain what we're doing here exactly?

File: notebooks/02-transformers.ipynb [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "## Want it lighter? Faster? Let's talk distillation ! \n",
+    "\n",
+    "One of the main concern when using this Transformer based models is the computational power they require. All over this notebook we are using BERT model as it can be run on common machines but that's not the case for all of the models.\n",
```
Comment (LysandreJik) [inline notebooks/02-transformers.ipynb]: concern > concerns

Comment (LysandreJik) [reply notebooks/02-transformers.ipynb]: this > these

File: notebooks/02-transformers.ipynb [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+    "\n",
+    "For example, Google released few months ago **T5** an Encoder/Decoder architecture based on Transformer and available in `transformers` with no more than 11 billions parameters. Microsoft also recently entered the game with **Turing-NLG** using 17 billions parameters. This kind of model require tens of gigabytes to store the weights and a tremendous compute infrastructure to run such models what makes it impracticable for the common man !\n",
```
Comment (LysandreJik) [inline notebooks/02-transformers.ipynb]: released few > released a few

Comment (LysandreJik) [reply notebooks/02-transformers.ipynb]: require tens > requires tens

Comment (LysandreJik) [reply notebooks/02-transformers.ipynb]: what > which

File: notebooks/02-transformers.ipynb [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+   "metadata": {},
+   "source": [
+    "## Community provided models\n",
+    "\n",
+    "Last but not least, earlier in this notebook we introduced Hugging Face `transformers` as a repository for the NLP community to exchange pretrained models. We wanted to highlight this features and all the possiblity it offers for the end-user.\n",
+    "\n",
+    "To leverage community pretrained models, just provide the organisation name and name of the model to `from_pretrained` and it will do all the magic for you ! \n",
+    "\n",
+    "\n",
+    "We currently have more 50 models provided by the community and more are added every day, don't hesitate to give it a try !"
```
Comment (LysandreJik) [inline notebooks/02-transformers.ipynb]: have more 50 > have more than 50

File: notebooks/03-pipelines.ipynb [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "Newly introduced in transformers v2.3.0, **pipelines** provides a high-level, easy-to-use,\n",
```
Comment (LysandreJik) [inline notebooks/03-pipelines.ipynb]: no comma between easy-to-use and API

File: notebooks/03-pipelines.ipynb [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+    "API for doing inference over a variety of downstream-tasks, including: \n",
+    "\n",
+    "- Sentence Classification (Sentiment Analysis): Indicate if the overall sentence is either positive or negative. _(Binary Classification task or Logitic Regression task)_\n",
+    "- Token Classification (Named Entity Recognition, Part-of-Speech tagging): For each sub-entities _(**tokens**)_ in the input, assign them a label _(Classification task)_.\n",
+    "- Question-Answering: Provided a tuple (question, context) the model should find the span of text in **content** answering the **question**.\n",
+    "- Mask-Filling: Generates the next word provided an initial **context**.\n",
```
Comment (LysandreJik) [inline notebooks/03-pipelines.ipynb]: Shouldn't it rather replace the masked token with a predicted token, rather than generating the next word?

File: notebooks/03-pipelines.ipynb [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "Alright ! Now you have a nice picture of what is possible through transformers' pipelines, and there is more\n",
+    "to come in future release. \n",
```
Comment (LysandreJik) [inline notebooks/03-pipelines.ipynb]: release > releases

**Review by LysandreJik — COMMENTED** [review]
Great, love it! Added a few comments, mostly regarding spelling and what I would change

File: notebooks/01-training-tokenizers.ipynb [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    "pycharm": {
+     "name": "#%% code\n",
+     "is_executing": false
+    }
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "Et voila ! You trained your very first tokenizer from scratch using `tokenizers`. Of course, this \n",
```
[the remark this answers is not in the document]
Comment (mfuntowicz) [reply notebooks/01-training-tokenizers.ipynb]: qwerty keyboard <3

**Review by LysandreJik — APPROVED** [review]
Great!
